tests/unit/s2n_self_talk_state_machine_test.c

```diff
@@ -0,0 +1,189 @@
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+#define BUFFER_SIZE 100
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    if (s2n_is_tls13_fully_supported()) {
+        /* Set up generic TLS13 config */
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(config);
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+        struct s2n_cert_chain_and_key *chain_and_key = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+
+        /* Generic TLS12 config */
+        DEFER_CLEANUP(struct s2n_config *tls12_config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(tls12_config);
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(tls12_config));
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(tls12_config, "default"));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(tls12_config, chain_and_key));
+
+        /* A connection cannot switch to the TLS12 state machine after the Server Hello */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+
+            /* Create nonblocking pipes */
+            struct s2n_test_io_pair io_pair;
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+            /* Do handshake until the cert message is reached */
+            EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn, SERVER_CERT));
+
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(client_conn), S2N_TLS13);
+
+            /* Alter which state machine the connection is using */
+            server_conn->actual_protocol_version = S2N_TLS12;
+            client_conn->actual_protocol_version = S2N_TLS12;
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn), S2N_ERR_STATE_MACHINE_SWITCH);
```

Comment:
I'd suggest calling s2n_negotiate separately for the client and server rather than relying on this method. s2n_negotiate_test_server_and_client will bail after the first S2N_ERR_STATE_MACHINE_SWITCH, so you're only testing the client or the server here, not both.